src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -5982,6 +5982,99 @@ MergeTreeData::MutableDataPartPtr MergeTreeData::cloneAndLoadDataPartOnSameDisk(
     return dst_data_part;
 }
 
+MergeTreeData::MutableDataPartPtr MergeTreeData::cloneAndLoadPartOnSameDiskWithDifferentPartitionKey(
+        const MergeTreeData::DataPartPtr & src_part, const String & tmp_part_prefix,
+        const MergeTreePartInfo & dst_part_info, const StorageMetadataPtr & metadata_snapshot,
+        const MergeTreePartition & new_partition, const IMergeTreeDataPart::MinMaxIndex & new_min_max_index,
+        const MergeTreeTransactionPtr & txn, HardlinkedFiles * hardlinked_files)
+{
+    /// Check that the storage policy contains the disk where the src_part is located.
+    bool does_storage_policy_allow_same_disk = false;
+    for (const DiskPtr & disk : getStoragePolicy()->getDisks())
+    {
+        if (disk->getName() == src_part->data_part_storage->getDiskName())
+        {
+            does_storage_policy_allow_same_disk = true;
+            break;
+        }
+    }
+
+    if (!does_storage_policy_allow_same_disk)
+        throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Could not clone and load part {} because disk does not belong to storage policy",
+                quoteString(src_part->data_part_storage->getFullPath()));
+
+    String dst_part_name = src_part->getNewName(dst_part_info);
+    assert(!tmp_part_prefix.empty());
+    String tmp_dst_part_name = tmp_part_prefix + dst_part_name;
+
+    /// Why it is needed if we only hardlink files?
+    auto reservation = src_part->data_part_storage->reserve(src_part->getBytesOnDisk());
+
+    auto src_part_storage = src_part->data_part_storage;
+
+    /// If source part is in memory, flush it to disk and clone it already in on-disk format
+    if (auto src_part_in_memory = asInMemoryPart(src_part))
+    {
+        auto flushed_part_path = src_part_in_memory->getRelativePathForPrefix(tmp_part_prefix);
+        src_part_storage = src_part_in_memory->flushToDisk(flushed_part_path, metadata_snapshot);
+    }
+
+    LOG_DEBUG(log, "Cloning part {} to {}",
+              src_part_storage->getFullPath(),
+              std::string(fs::path(src_part_storage->getFullRootPath()) / tmp_dst_part_name));
+
+    auto dst_part_storage = src_part_storage->freeze(relative_data_path, tmp_dst_part_name, /* make_source_readonly */ false, {}, /* copy_instead_of_hardlinks */ false);
+
+    auto dst_data_part = createPart(dst_part_name, dst_part_info, dst_part_storage);
+
+    dst_data_part->minmax_idx->replace(new_min_max_index);
+
+    auto volume = getStoragePolicy()->getVolume(0);
+
+    auto data_part_storage_builder = std::make_shared<DataPartStorageBuilderOnDisk>(volume, getRelativeDataPath(), tmp_dst_part_name);
```

Comment:
Is this correct? Is there a ready-to-use instance of data_part_storage_builder in order to avoid instantiating it?